### PR TITLE
Commit changes in module 4

### PR DESCRIPTION
**Pull request by adrianadiaz02**

### CODE REVIEW

File: src/module_4/push_model.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class PushModel:
+    FEATURE_COLUMNS = ["user_order_seq", "ordered_before", "abandoned_before", 
+        "active_snoozed", "set_as_regular", "normalised_price", "discount_pct", 
+        "global_popularity", "count_adults", "count_children", "count_babies", "count_pets", 
+        "people_ex_baby", "days_since_purchase_variant_id", "avg_days_to_buy_variant_id", 
+        "std_days_to_buy_variant_id", "days_since_purchase_product_type", 
+        "avg_days_to_buy_product_type", "std_days_to_buy_product_type"]
+        
+    TARGET_COLUMN = "outcome"
+
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:12]: I miss type hinting and a docstring indicating what this method expects. Otherwise, it is hard for any of your colleagues to understand how to use

File: src/module_4/push_model.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        - a series with the target column"""
+        X = self.extract_features(df)
+        Y = self.extract_target(df)
+        return X, Y
+
+
+    def fit_model(self, df):
+        """Given a complete dataframe (with both features and target columns), fits a Gradient Boosting model"""
+        features, label = self.feature_target_split(df)
+        self.gbc.fit(features, label)
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:39]: Why do you return the inner model?
If you build a class is for users to interact directly with it, not to use it as subroutines

File: src/module_4/push_model.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    def feature_target_split(self, df):
+        """Given a dataframe, divides the input in two subsets:
+        - a dataframe with the features subset
+        - a series with the target column"""
+        X = self.extract_features(df)
+        Y = self.extract_target(df)
+        return X, Y
+
+
+    def fit_model(self, df):
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:36]: I miss a predict method here

File: src/module_4/push_model.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+from sklearn.ensemble import GradientBoostingClassifier
```
Comment (pabaldonedo) [inline src/module_4/push_model.py:1]: Mostly good. However, if you build a class is to make users interact fully with it and not as a helper to get a sklearn model as a return as it happens in fit method.

Think that building a class helps you building an interface: whoever wants to use the push model call this class.
In the future, you will then be able to modify the internals fully without breakung any pipeline as long you keep the same interface

File: src/module_4/utils.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+from pathlib import Path
+import pandas as pd
+import logging
+
+logger = logging.getLogger(__name__)
+logger.level = logging.INFO
+
+def load_raw_dataset():
```
Comment (pabaldonedo) [inline src/module_4/utils.py:8]: It is a good practice to get use to always type hint

File: src/module_4/utils.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    file_path = Path("/mnt/c/Users/Adriana/Desktop/ZRIVE/data/groceries/box_builder_dataset/feature_frame.csv")
+    data = pd.read_csv(file_path)
+    logger.info(f"Dataset loaded from {file_path}")
+    return data
+
```
Comment (pabaldonedo) [inline src/module_4/utils.py:13]: Being picky: try to follow PEP8 style guide. Here two blank lines are expecred

File: src/module_4/utils.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
+from pathlib import Path
+import pandas as pd
+import logging
+
+logger = logging.getLogger(__name__)
+logger.level = logging.INFO
+
+def load_raw_dataset():
+    file_path = Path("/mnt/c/Users/Adriana/Desktop/ZRIVE/data/groceries/box_builder_dataset/feature_frame.csv")
```
Comment (pabaldonedo) [inline src/module_4/utils.py:9]: You could let this as input parameter wirh default value so this method is more reusable.
Think always ahead of the future uses. It saves time in the long run

File: src/module_4/utils.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
+from pathlib import Path
```
Comment (pabaldonedo) [inline src/module_4/utils.py:1]: Docstrings and type hinting would have been helpful

File: src/module_4/predict.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
+import json
+import pandas as pd
+import joblib
+
+class PushModel:
```
Comment (pabaldonedo) [inline src/module_4/predict.py:5]: You must import the PushModel from the pus model file and there implemenr all the require logic. With that you will have everything centralised and easier to maintain

File: src/module_4/predict.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+    """
+
+    model_path = event.get("model_path", "push_latest.pkl")
+    users_data = json.loads(event["users"])
+    
+    data_to_predict = pd.DataFrame.from_dict(users_data, orient='index')
+    
+    model = PushModel(model_path)
+    predictions = model.predict(data_to_predict)
```
Comment (pabaldonedo) [inline src/module_4/predict.py:31]: What if the input data does not contain the necessary fields?
Checking and returning informed error early is good practice.

For this you make use of pydantic package. You may come accross it on Sergio Rozada's lecture

File: src/module_4/predict.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    return {
+        "statusCode": 200,
+        "body": json.dumps({
+            "prediction": predictions_dict
+        })
+    }
+
+
+# Example event
+event = {
```
Comment (pabaldonedo) [inline src/module_4/predict.py:44]: This will be executed whenever you import the package.

If it is just an example you can add a docstring at the top of the module

File: src/module_4/fit.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    return {
+        "statusCode": 200,
+        "body": json.dumps({
+            "model_path": model_path
+        })
+    }
+
+
+# Example event
+event = {
```
Comment (pabaldonedo) [inline src/module_4/fit.py:48]: This should go to a docstring

File: src/module_4/utils.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+logger = logging.getLogger(__name__)
+logger.level = logging.INFO
+
+def load_raw_dataset():
+    file_path = Path("/mnt/c/Users/Adriana/Desktop/ZRIVE/data/groceries/box_builder_dataset/feature_frame.csv")
+    data = pd.read_csv(file_path)
+    logger.info(f"Dataset loaded from {file_path}")
+    return data
+
+def correct_data_types(data):
```
Comment (pabaldonedo) [inline src/module_4/utils.py:14]: You should be very explicit that this is changing data outside the scope of the method (you are changind the dataframe in place)

Generally this can be sources of bugs. But, when it make sense to do it, be at least very explicit in the docstring

File: src/module_4/Module_4.md [hunk 14]
```diff
@@ -0,0 +1,1998 @@
...
+        # Convert the column to integer type
+        data[col] = data[col].astype(int)
+
+data['people_ex_baby'] = data['people_ex_baby'].astype('int64')
+
+data['days_since_purchase_variant_id'] = data['days_since_purchase_variant_id'].astype('int64')
+data['days_since_purchase_product_type'] = data['days_since_purchase_product_type'].astype('int64')
+```
+
+It is worth mentioning that, since we have being working on this dataset over the latest several weeks, this project skips the Exploratory Data Analysis (EDA) phase as it is already available in previous reports. If you want to have a look at the EDA performed, please check these two files:
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:347]: Very good to redirect to past analysis for new readers. If there were some very relevant facts you may repeat here in a bullet point list

File: src/module_4/Module_4.md [hunk 15]
```diff
@@ -0,0 +1,1998 @@
...
+        fpr, tpr, _ = roc_curve(y_true, y_pred)
+        roc_auc = auc(fpr, tpr)
+
+        axes[1].plot(fpr, tpr)
+        axes[1].set_xlabel('False Positive Rate')
+        axes[1].set_ylabel('True Positive Rate')
+        axes[1].set_title(f'ROC Curve (AUC={roc_auc:.2f})')
+
+    plt.tight_layout()
+    plt.show()
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:497]: This is a matter of persnal taste, but another option is:

- return the fig and ax and pass the responsability of showing to the caller

Why? You allow to expand the fig easily if required in a specific case. Contrary you may argue that this is a notebook and it is good enough as it is

File: src/module_4/Module_4.md [hunk 16]
```diff
@@ -0,0 +1,1998 @@
...
+
+As we can observe, this model provided an AUC of the Precision-Recall curve of 0.16 and 0.14 for the training and validation sets respectively. On the other hand, the AUC of the ROC curve had a value of 0.81 for both subsets.
+
+### 4. Non-linear Models
+We will now try to improve our model by increasing its complexity.
+
+#### 4.1 Random Forest
+In a Random Forest, each tree receives a subset of the data, and each split only looks at a random subset of the features. Additionaly, the trees are constructed until the end and the decision is made by a voting process. Furthermore, it is worth mentioning that is hass three hyperparameters: the number of trees, the variables at each split, and the sample size at each tree. Nevertheless, we will only define the number of trees, and set the other hyperparameters with their default value.
+
+It is also worth mentioning that now we do not need to scale the data, as there is no need to add regularisation.
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:574]: These detail descriptions on the model basics are nice here since this is a training. But just reminding that at the workplace these are unneeded and add clutter to the text

File: src/module_4/Module_4.md [hunk 17]
```diff
@@ -0,0 +1,1998 @@
...
+    For 100 trees:
+    Train AUC: 0.9939, Validation AUC: 0.7387
+    Train Cross Entropy: 0.0238, Validation Cross Entropy: 0.1829
+    Train Average Precision: 0.8090, Validation Average Precision: 0.0919
+    
+
+
+Once again, we can observe that as the number of trees increases, the model's performance on the validation set consistently improves. Increasing the number of trees results in progressively better generalization, as indicated by rising validation AUC and Average Precision scores, and decreasing validation Cross Entropy loss. 
+
+We also have still high variance due to the fact that there is correlation between features, and therefore correlation between trees. Additionally, we don't have an AUC of exactly 100%, as there are samples with same input but different output, which indicates that there is either noise or we need more features (additional features may make previous equal instances different).
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:1145]: This is only an intuition. We have not carried out an analysis to categorically state it

File: src/module_4/Module_4.md [hunk 18]
```diff
@@ -0,0 +1,1998 @@
...
+```
+
+
+    
+![png](Module_4_files/Module_4_58_0.png)
+    
+
+
+
+    The Kernel crashed while executing code in the current cell or a previous cell.
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:1982]: Have you checked why this happened?

File: src/module_4/Module_4.md [hunk 19]
```diff
@@ -0,0 +1,1998 @@
+# Module 4: TDD
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:1]: Good work. I appreciate the detail comments. If you have time, I would strongly encourage you to explore more things than presented in class for all zrive modules. I felt this was too tightly related to my solution. Which is good to go over it again but even better if go also to new places on your own

File: src/module_4/Module_4.md [hunk 20]
```diff
@@ -0,0 +1,1998 @@
...
+
+
+    View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details.
+
+
+We can osberve that in the Precision-Recall Curve, the Logistic Regression has the highest average precision (AP),compared to Random Forest and Gradient Boosting. Furthermore, Random Forest has a moderate AP, while Gradient Boosting has the lowest AP, indicating weaker performance in distinguishing the positive class.
+
+On the other hand, in the ROC Curve, Logistic Regression and Gradient Boosting show similar performance with AUCs of 0.81 and 0.80, respectively. Random Forest performs worse with an AUC of 0.73, indicating a higher rate of false positives and/or false negatives compared to the other two models.
+
+In conclusion, Logistic Regression is the best performer in terms of both Precision-Recall and ROC metrics; Gradient Boosting has comparable ROC performance to Logistic Regression but performs worse in the Precision-Recall metric; and, finally, Random Forest has the lowest performance in both Precision-Recall and ROC metrics.
```
Comment (pabaldonedo) [inline src/module_4/Module_4.md:1998]: It looks odd the difference between val and test. Either there is heavy data shift or bug. This is worth investigating

**Review by pabaldonedo — APPROVED** [review]
Good work. I have left some comments on the way. My 3 main takeaways:

 - If you have time try to explore new things on your own. I felt the work was super tight to my solution. If you do not have much time it is ok, since it allows you to work on the material. But if you do new things I can add more value for you while reviewing too.

- Model class is meant to encapsulate everything and serve as an inteeface ypu interact with.

- The drop in test performance vs validation looks odd. Either there are bugs or heavy data shift. Further investigation must be taken


Keep up the good work!
